File: Runtime/Plugins/unity.androidlib/src/main/java/com/affise/attribution/unity/AffiseWrapper.java
package com.affise.attribution.unity;

import android.app.Application;
import android.content.Context;

import com.affise.attribution.Affise;
import com.affise.attribution.events.Event;
import com.affise.attribution.events.autoCatchingClick.AutoCatchingType;
import com.affise.attribution.init.AffiseInitProperties;
import com.affise.attribution.unity.common.NativeEventCallback;
import com.affise.attribution.unity.common.Result;
import com.affise.attribution.unity.ext.AffiseInitPropertiesExt;
import com.affise.attribution.unity.ext.AutoCatchingTypeExt;
import com.affise.attribution.unity.factories.AffiseEvensFactory;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class AffiseWrapper {

    private static final String AFFISE_INIT = "init";
    private static final String AFFISE_SEND_EVENTS = "send_events";
    private static final String AFFISE_SEND_EVENT = "send_event";
    private static final String AFFISE_ADD_PUSH_TOKEN = "add_push_token";
    private static final String AFFISE_REGISTER_WEB_VIEW = "register_web_view";
    private static final String AFFISE_UNREGISTER_WEB_VIEW = "unregister_web_view";
    private static final String AFFISE_REGISTER_DEEPLINK_CALLBACK = "register_deeplink_callback";
    private static final String AFFISE_SET_SECRET_ID = "set_secret_id";
    private static final String AFFISE_SET_AUTO_CATCHING_TYPES = "set_auto_catching_types";
    private static final String AFFISE_SET_OFFLINE_MODE_ENABLED = "set_offline_mode_enabled";
    private static final String AFFISE_IS_OFFLINE_MODE_ENABLED = "is_offline_mode_enabled";
    private static final String AFFISE_SET_BACKGROUND_TRACKING_ENABLED = "set_background_tracking_enabled";
    private static final String AFFISE_IS_BACKGROUND_TRACKING_ENABLED = "is_background_tracking_enabled";
    private static final String AFFISE_SET_TRACKING_ENABLED = "set_tracking_enabled";
    private static final String AFFISE_IS_TRACKING_ENABLED = "is_tracking_enabled";
    private static final String AFFISE_FORGET = "forget";
    private static final String AFFISE_SET_ENABLED_METRICS = "set_enabled_metrics";
    private static final String AFFISE_CRASH_APPLICATION = "crash_application";
    private static final String AFFISE_GET_REFERRER = "get_referrer";

    private final Context applicationContext;
    private final AffiseEvensFactory evensFactory;
    private NativeEventCallback eventCallback;

    public AffiseWrapper(Context applicationContext) {
        this.applicationContext = applicationContext;
        evensFactory = new AffiseEvensFactory();
    }

    public void setEventCallback(NativeEventCallback eventCallback) {
        this.eventCallback = eventCallback;
    }

    public void handle(String methodName, JSONObject json, Result result) {
        switch (methodName) {
            case AFFISE_INIT:
                nativeInit(json, result);
                break;
            case AFFISE_SEND_EVENTS:
                nativeSendEvents(json, result);
                break;
            case AFFISE_SEND_EVENT:
                nativeSendEvent(json, result);
                break;
            case AFFISE_ADD_PUSH_TOKEN:
                nativeAddPushToken(json, result);
                break;
            case AFFISE_REGISTER_WEB_VIEW:
                nativeRegisterWebView(json, result);
                break;
            case AFFISE_UNREGISTER_WEB_VIEW:
                nativeUnregisterWebView(json, result);
                break;
            case AFFISE_REGISTER_DEEPLINK_CALLBACK:
                nativeRegisterDeeplinkCallback(json, result);
                break;
            case AFFISE_SET_SECRET_ID:
                nativeSetSecretId(json, result);
                break;
            case AFFISE_SET_AUTO_CATCHING_TYPES:
                nativeSetAutoCatchingTypes(json, result);
                break;
            case AFFISE_SET_OFFLINE_MODE_ENABLED:
                nativeSetOfflineModeEnabled(json, result);
                break;
            case AFFISE_IS_OFFLINE_MODE_ENABLED:
                nativeIsOfflineModeEnabled(json, result);
                break;
            case AFFISE_SET_BACKGROUND_TRACKING_ENABLED:
                nativeSetBackgroundTrackingEnabled(json, result);
                break;
            case AFFISE_IS_BACKGROUND_TRACKING_ENABLED:
                nativeIsBackgroundTrackingEnabled(json, result);
                break;
            case AFFISE_SET_TRACKING_ENABLED:
                nativeSetTrackingEnabled(json, result);
                break;
            case AFFISE_IS_TRACKING_ENABLED:
                nativeIsTrackingEnabled(json, result);
                break;
            case AFFISE_FORGET:
                nativeForget(json, result);
                break;
            case AFFISE_SET_ENABLED_METRICS:
                nativeSetEnabledMetrics(json, result);
                break;
            case AFFISE_CRASH_APPLICATION:
                nativeCrashApplication(json, result);
                break;
            case AFFISE_GET_REFERRER:
                nativeGetReferrer(json, result);
                break;
        }
    }

    private void nativeInit(JSONObject json, Result result) {
        Application app = (Application) applicationContext;
        if (app == null) {
            result.error("No application context", null, null);
            return;
        }
        AffiseInitProperties properties = AffiseInitPropertiesExt.toAffiseInitProperties(json);
        Affise._crossPlatform.unity();
        Affise.init(app, properties);
        Affise._crossPlatform.start();
    }

    private void nativeSendEvents(JSONObject json, Result result) {
        Affise.sendEvents();
    }

    private void nativeSendEvent(JSONObject json, Result result) {
        Event event = evensFactory.event(json);
        if (event == null) {
            result.error("Error event factory", null, null);
            return;
        }
        Affise.sendEvent(event);
    }

    private void nativeAddPushToken(JSONObject json, Result result) {
        String token = json.optString(AFFISE_ADD_PUSH_TOKEN);
        if (token.isEmpty()) {
            result.error("Error token is empty", null, null);
            return;
        }
        Affise.addPushToken(token);
    }

    private void nativeRegisterWebView(JSONObject json, Result result) {
        result.notImplemented();
    }

    private void nativeUnregisterWebView(JSONObject json, Result result) {
        result.notImplemented();
    }

    private void nativeRegisterDeeplinkCallback(JSONObject json, Result result) {
        Affise.registerDeeplinkCallback(uri -> {
            if (this.eventCallback != null) {
                this.eventCallback.HandleEvent(AFFISE_REGISTER_DEEPLINK_CALLBACK, uri.toString());
                return true;
            }
            return false;
        });
    }

    private void nativeSetSecretId(JSONObject json, Result result) {
        String secretId = json.optString(AFFISE_SET_SECRET_ID);
        if (secretId.isEmpty()) {
            result.error("Error secret id is empty", null, null);
            return;
        }
        Affise.setSecretId(secretId);
    }

    private void nativeSetAutoCatchingTypes(JSONObject json, Result result) {
        JSONArray data = json.optJSONArray(AFFISE_SET_AUTO_CATCHING_TYPES);
        if (data == null) {
            result.error("Error AutoCatchingTypes not set", null, null);
            return;
        }
        List<AutoCatchingType> types = new ArrayList<>();
        AutoCatchingType type;
        for (int i = 0; i < data.length(); i++) {
            type = AutoCatchingTypeExt.toAutoCatchingType(data.optString(i));
            if (type == null) continue;
            types.add(type);
        }

        Affise.setAutoCatchingTypes(types);
    }

    private void nativeSetOfflineModeEnabled(JSONObject json, Result result) {
        if (!json.has(AFFISE_SET_OFFLINE_MODE_ENABLED)) {
            result.error("Error offline mode value not set", null, null);
            return;
        }
        boolean enabled = json.optBoolean(AFFISE_SET_OFFLINE_MODE_ENABLED);
        Affise.setOfflineModeEnabled(enabled);
    }

    private void nativeIsOfflineModeEnabled(JSONObject json, Result result) {
        result.success(Affise.isOfflineModeEnabled());
    }

    private void nativeSetBackgroundTrackingEnabled(JSONObject json, Result result) {
        if (!json.has(AFFISE_SET_BACKGROUND_TRACKING_ENABLED)) {
            result.error("Error background tracking value not set", null, null);
            return;
        }
        boolean enabled = json.optBoolean(AFFISE_SET_BACKGROUND_TRACKING_ENABLED);
        Affise.setBackgroundTrackingEnabled(enabled);
    }

    private void nativeIsBackgroundTrackingEnabled(JSONObject json, Result result) {
        result.success(Affise.isBackgroundTrackingEnabled());
    }

    private void nativeSetTrackingEnabled(JSONObject json, Result result) {
        if (!json.has(AFFISE_SET_TRACKING_ENABLED)) {
            result.error("Error tracking value not set", null, null);
            return;
        }
        boolean enabled = json.optBoolean(AFFISE_SET_TRACKING_ENABLED);
        Affise.setTrackingEnabled(enabled);
    }

    private void nativeIsTrackingEnabled(JSONObject json, Result result) {
        result.success(Affise.isTrackingEnabled());
    }

    private void nativeForget(JSONObject json, Result result) {
        String userData = json.optString(AFFISE_FORGET);
        if (userData.isEmpty()) {
            result.error("Error forget userData is empty", null, null);
            return;
        }
        Affise.forget(userData);
    }

    private void nativeSetEnabledMetrics(JSONObject json, Result result) {
        if (!json.has(AFFISE_SET_ENABLED_METRICS)) {
            result.error("Error enabled metrics value not set", null, null);
            return;
        }
        boolean enabled = json.optBoolean(AFFISE_SET_ENABLED_METRICS);
        Affise.setEnabledMetrics(enabled);
    }

    private void nativeCrashApplication(JSONObject json, Result result) {
        Affise.crashApplication();
    }

    private void nativeGetReferrer(JSONObject json, Result result) {
        Affise.getReferrer(referrer -> {
            if (this.eventCallback != null) {
                this.eventCallback.HandleEvent(AFFISE_GET_REFERRER, referrer);
            }
        });
    }
}

File: Runtime/Plugins/iOS/Source/AffiseWrapper.swift
import AffiseAttributionLib


internal class AffiseWrapper {

    private let AFFISE_INIT = "init"
    private let AFFISE_SEND_EVENTS = "send_events"
    private let AFFISE_SEND_EVENT = "send_event"
    private let AFFISE_ADD_PUSH_TOKEN = "add_push_token"
    private let AFFISE_REGISTER_WEB_VIEW = "register_web_view"
    private let AFFISE_UNREGISTER_WEB_VIEW = "unregister_web_view"
    private let AFFISE_REGISTER_DEEPLINK_CALLBACK = "register_Deeplink_Callback"
    private let AFFISE_SET_SECRET_ID = "set_secret_id"
    private let AFFISE_SET_AUTO_CATCHING_TYPES = "set_auto_catching_types"
    private let AFFISE_SET_OFFLINE_MODE_ENABLED = "set_offline_mode_enabled"
    private let AFFISE_IS_OFFLINE_MODE_ENABLED = "is_offline_mode_enabled"
    private let AFFISE_SET_BACKGROUND_TRACKING_ENABLED = "set_background_tracking_enabled"
    private let AFFISE_IS_BACKGROUND_TRACKING_ENABLED = "is_background_tracking_enabled"
    private let AFFISE_SET_TRACKING_ENABLED = "set_tracking_enabled"
    private let AFFISE_IS_TRACKING_ENABLED = "is_tracking_enabled"
    private let AFFISE_FORGET = "forget"
    private let AFFISE_SET_ENABLED_METRICS = "set_enabled_metrics"
    private let AFFISE_CRASH_APPLICATION = "crash_application"
    private let AFFISE_GET_REFERRER = "get_referrer"
    private let AFFISE_SKAD_REGISTER = "skad_register"
    private let AFFISE_SKAD_POSTBACK = "skad_postback"

    private let AFFISE_HANDLE_INITIAL_LINK = "handle_initial_link"

    private let UNITY_SKAD_REGISTER_ERROR = "skadRegisterError"
    private let UNITY_SKAD_POSTBACK_ERROR = "skadPostbackError"
    
    private let evensFactory: AffiseEvensFactory = AffiseEvensFactory()

    var application: UIApplication?
    var launchOptions: [UIApplication.LaunchOptionsKey: Any]?
    var deepLink: String?
    
    private var sendEvent: ((String, String)->Void)? = nil
    
    public func SetEventCallback(_ callback: @escaping (String, String)->Void) {
        sendEvent = callback
    }
    
    public func handle(method: String, data: [String: Any?], result: NativeResult) {
        switch method {
        case AFFISE_INIT: nativeInit(data: data, result: result)
        case AFFISE_SEND_EVENTS: nativeSendEvents(data: data, result: result)
        case AFFISE_SEND_EVENT: nativeSendEvent(data: data, result: result)
        case AFFISE_ADD_PUSH_TOKEN: nativeAddPushToken(data: data, result: result)
        case AFFISE_REGISTER_WEB_VIEW: nativeRegisterWebView(data: data, result: result)
        case AFFISE_UNREGISTER_WEB_VIEW: nativeUnregisterWebView(data: data, result: result)
        case AFFISE_REGISTER_DEEPLINK_CALLBACK: nativeRegisterDeeplinkCallback(data: data, result: result)
        case AFFISE_SET_SECRET_ID: nativeSetSecretId(data: data, result: result)
        case AFFISE_SET_AUTO_CATCHING_TYPES: nativeSetAutoCatchingTypes(data: data, result: result)
        case AFFISE_SET_OFFLINE_MODE_ENABLED: nativeSetOfflineModeEnabled(data: data, result: result)
        case AFFISE_IS_OFFLINE_MODE_ENABLED: nativeIsOfflineModeEnabled(data: data, result: result)
        case AFFISE_SET_BACKGROUND_TRACKING_ENABLED: nativeSetBackgroundTrackingEnabled(data: data, result: result)
        case AFFISE_IS_BACKGROUND_TRACKING_ENABLED: nativeIsBackgroundTrackingEnabled(data: data, result: result)
        case AFFISE_SET_TRACKING_ENABLED: nativeSetTrackingEnabled(data: data, result: result)
        case AFFISE_IS_TRACKING_ENABLED: nativeIsTrackingEnabled(data: data, result: result)
        case AFFISE_FORGET: nativeForget(data: data, result: result)
        case AFFISE_SET_ENABLED_METRICS: nativeSetEnabledMetrics(data: data, result: result)
        case AFFISE_CRASH_APPLICATION: nativeCrashApplication(data: data, result: result)
        case AFFISE_GET_REFERRER: nativeGetReferrer(data: data, result: result)
        case AFFISE_HANDLE_INITIAL_LINK: nativeHandleInitialLink(data: data, result: result)
        case AFFISE_SKAD_REGISTER: nativeSkadRegister(data: data, result: result)
        case AFFISE_SKAD_POSTBACK: nativeSkadPostback(data: data, result: result)
        default:
            result.notImplemented()
        }
    }
    
    private func nativeInit(data: [String: Any?], result: NativeResult) {        
        if application == nil {
            result.error("No application context")
            return
        }

        Affise._crossPlatform.unity()
        Affise.shared.load(app: application!, initProperties: data.toAffiseInitProperties , launchOptions: launchOptions)
    }
    
    private func nativeSendEvent(data: [String: Any?], result: NativeResult) {
        if let event = evensFactory.event(data) {
            Affise.shared.sendEvent(event: event)
        } else {
            result.error("Wrong argument map")
        }
    }
    
    private func nativeSendEvents(data: [String: Any?], result: NativeResult) {
        Affise.shared.sendEvents()
    }
    
    private func nativeAddPushToken(data: [String: Any?], result: NativeResult) {
        if let pushToken = data[AFFISE_ADD_PUSH_TOKEN] as? String {
            Affise.shared.addPushToken(pushToken: pushToken)
        }
    }
    
    private func nativeRegisterWebView(data: [String: Any?], result: NativeResult) {
        // TODO registerWebView
        // Affise.shared.registerWebView(webView)
        result.notImplemented()
    }
    
    private func nativeUnregisterWebView(data: [String: Any?], result: NativeResult) {
        Affise.shared.unregisterWebView()
    }
    
    private func nativeRegisterDeeplinkCallback(data: [String: Any?], result: NativeResult) {
        registerCallback()
    }
    
    private func nativeSetSecretId(data: [String: Any?], result: NativeResult) {
        if let secretId = data[AFFISE_SET_SECRET_ID] as? String {
            Affise.shared.setSecretId(secretId: secretId)
        }
    }
    
    private func nativeSetAutoCatchingTypes(data: [String: Any?], result: NativeResult) {
        result.notImplemented()
    }
    
    private func nativeSetOfflineModeEnabled(data: [String: Any?], result: NativeResult) {
        let enabled = data[AFFISE_SET_OFFLINE_MODE_ENABLED] as? Bool ?? false
        Affise.shared.setOfflineModeEnabled(enabled: enabled)
    }
    
    private func nativeIsOfflineModeEnabled(data: [String: Any?], result: NativeResult) {
        result.success(Affise.shared.isOfflineModeEnabled())
    }
    
    private func nativeSetBackgroundTrackingEnabled(data: [String: Any?], result: NativeResult) {
        let enabled = data[AFFISE_SET_BACKGROUND_TRACKING_ENABLED] as? Bool ?? false
        Affise.shared.setBackgroundTrackingEnabled(enabled: enabled)
    }
    
    private func nativeIsBackgroundTrackingEnabled(data: [String: Any?], result: NativeResult) {
        result.success(Affise.shared.isBackgroundTrackingEnabled())
    }
    
    private func nativeSetTrackingEnabled(data: [String: Any?], result: NativeResult) {
        let enabled = data[AFFISE_SET_TRACKING_ENABLED] as? Bool ?? false
        Affise.shared.setTrackingEnabled(enabled: enabled)
    }
    
    private func nativeIsTrackingEnabled(data: [String: Any?], result: NativeResult) {
        result.success(Affise.shared.isTrackingEnabled())
    }
    
    private func nativeForget(data: [String: Any?], result: NativeResult) {
        result.notImplemented()
    }
    
    private func nativeSetEnabledMetrics(data: [String: Any?], result: NativeResult) {
        result.notImplemented()
    }
    
    private func nativeCrashApplication(data: [String: Any?], result: NativeResult) {
        result.notImplemented()
    }
    
    private func nativeGetReferrer(data: [String: Any?], result: NativeResult) {
        result.notImplemented()
    }

    private func nativeHandleInitialLink(data: [String: Any?], result: NativeResult) {
        nativeHandleDeeplink(deepLink)
    }

    private func nativeSkadRegister(data: [String: Any?], result: NativeResult) {
        result.notImplemented()
    }

    private func nativeSkadPostback(data: [String: Any?], result: NativeResult) {
        result.notImplemented()
    }

    private func registerCallback() {
        Affise.shared.registerDeeplinkCallback {url in
        }
    }

    func nativeHandleDeeplink(_ url: String?) {
        if let url = url, !url.isEmpty {
            Affise._crossPlatform.handleDeeplink(uri: url)
            sendEvent?(AFFISE_REGISTER_DEEPLINK_CALLBACK, url)
        }
    }
}
